### ISSUE TITLE
EADDRINUSE on connect

### ISSUE THREAD

**Issue by konsumer** [request/request#210]
I am trying to write a little script to replicate an elasticsearch index to another elasticsearch index.

Basically, I need a HTTP connection to a source db, and to target db. I thought that my connections would be pooled automatically, but after a short time of running this code on 200,000 records I get a 

{ [Error: connect EADDRINUSE] code: 'EADDRINUSE', errno: 'EADDRINUSE', syscall: 'connect' }

Here are the important bits of my code:
https://gist.github.com/2128623

What am I doing wrong?

**Comment by tomhalley89** [comment]
Did anyone ever find a better work around for this? I'm currently sending a huge amount of requests to a node js server with request, and it too dies after about 16000 records. I'm working on windows, and after it times out at the 16000 mark, it takes approximately 30 seconds before it starts responding to requests again, even if i reboot the server. 

[It seems this is related](https://groups.google.com/forum/#!topic/nodejs/3KNosjnfTks)

**Comment by tomhalley89** [comment]
[And this also seems to be related](https://zeromq.jira.com/browse/LIBZMQ-519)

I moved my node script that was acting as a client, as well as the server onto a linux virtual machine, and from there i seems to work fine. I can make any number of requests and this won't happen, confirming this is a windows issue. 

I've also tested running the server on the linux virtual box and having the client running from my host machine, and it seemed to work without stopping at 16,000.

In summary, to anyone else who finds this issue through google, it's Windows. You will need to run your server from a Linux box. If you're testing a locally running server, I recommend using Vagrant to create a Linux box and run the server off of that.

**Comment by hassy** [comment]
This is most likely due to Windows running out of ephemeral ports.

To fix, set `MaxUserPort` to something like 65000 under `HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\ Services\TCPIP\Parameters `

Source: https://docs.oracle.com/cd/E26180_01/Search.94/ATGSearchAdmin/html/s1207adjustingtcpsettingsforheavyload01.html

**Comment by danielstaleiny** [comment]
I think another option is async library where you can select to wait for request. 
`async.eachLimit(collection,limit,function(item, callback){},function(errHandler(err)){})`

**Comment by ericrini** [comment]
Look at this thread for more ideas.

http://stackoverflow.com/questions/15909884/sockets-dont-appear-to-be-closing-when-using-node-js-http-get
